CDI Portable Extension QS

LightGuard opened the pull request:
Code complete for cdi-portable-extension

pom, template readme, beans.xml files missed by IntelliJ

Adding openshift profile

README done

LightGuard committed:
CDI Portable Extension Quickstart
Code complete for cdi-portable-extension

pom, template readme, beans.xml files missed by IntelliJ

Adding openshift profile

README done

pmuir commented on cdi-portable-extension/pom.xml:
```diff
@@ -0,0 +1,197 @@
+                </dependency>
+            </dependencies>
+        </profile>
+        <profile>
```
Do you really want this profile? You don't define OpenShift support in the README.

LightGuard replied:
Was simply copying what we had in others. If we don't see a point in having
remote we could drop it. Tasks uses remote as well.

On Thu, Aug 2, 2012 at 4:10 PM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> > -            </dependencies>
> > -        </profile>
> >   +
> > -        <profile>
> > -            <!-- An optional Arquillian testing profile that executes
> >   tests in a remote JBoss AS instance -->
> > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > -            <id>arq-jbossas-remote</id>
> > -            <dependencies>
> > -                <dependency>
> > -                    <groupId>org.jboss.as</groupId>
> >   +
> >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > -                    <scope>test</scope>
> > -                </dependency>
> > -            </dependencies>
> > -        </profile>
> > -        <profile>
> 
> Do you really want this profile?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297961

## 

Jason Porter
http://lightguard-jp.blogspot.com
http://twitter.com/lightguardjp

Software Engineer
Open Source Advocate
Author of Seam Catch - Next Generation Java Exception Handling

PGP key id: 926CCFF5
PGP key available at: keyserver.net, pgp.mit.edu

pmuir replied:
I meant the openshift profile :-)

LightGuard replied:
Ah, the openshift one. I initially didn't see the point, but we are trying
to push openshift, aren't we?

On Thu, Aug 2, 2012 at 4:15 PM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> > -            </dependencies>
> > -        </profile>
> >   +
> > -        <profile>
> > -            <!-- An optional Arquillian testing profile that executes
> >   tests in a remote JBoss AS instance -->
> > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > -            <id>arq-jbossas-remote</id>
> > -            <dependencies>
> > -                <dependency>
> > -                    <groupId>org.jboss.as</groupId>
> >   +
> >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > -                    <scope>test</scope>
> > -                </dependency>
> > -            </dependencies>
> > -        </profile>
> > -        <profile>
> 
> I meant the openshift profile :-)
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297983

## 

Jason Porter
http://lightguard-jp.blogspot.com
http://twitter.com/lightguardjp

Software Engineer
Open Source Advocate
Author of Seam Catch - Next Generation Java Exception Handling

PGP key id: 926CCFF5
PGP key available at: keyserver.net, pgp.mit.edu

pmuir replied:
Well, how would this qs run on openshift?

On 2 Aug 2012, at 23:18, Jason Porter wrote:

> > -            </dependencies>
> > -        </profile>
> >   +
> > -        <profile>
> > -            <!-- An optional Arquillian testing profile that executes tests in a remote JBoss AS instance -->
> > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > -            <id>arq-jbossas-remote</id>
> > -            <dependencies>
> > -                <dependency>
> > -                    <groupId>org.jboss.as</groupId>
> > -                    <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > -                    <scope>test</scope>
> > -                </dependency>
> > -            </dependencies>
> > -        </profile>
> > -        <profile>
> 
> Ah, the openshift one. I initially didn't see the point, but we are trying
> to push openshift, aren't we?
> 
> On Thu, Aug 2, 2012 at 4:15 PM, Pete Muir <
> reply@reply.github.com
> 
> > wrote:
> > 
> > > -            </dependencies>
> > > -        </profile>
> > >   +
> > > -        <profile>
> > > -            <!-- An optional Arquillian testing profile that executes
> > >   tests in a remote JBoss AS instance -->
> > > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > > -            <id>arq-jbossas-remote</id>
> > > -            <dependencies>
> > > -                <dependency>
> > > -                    <groupId>org.jboss.as</groupId>
> > >   +
> > >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > > -                    <scope>test</scope>
> > > -                </dependency>
> > > -            </dependencies>
> > > -        </profile>
> > > -        <profile>
> > 
> > I meant the openshift profile :-)
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297983
> 
> ## 
> 
> Jason Porter
> http://lightguard-jp.blogspot.com
> http://twitter.com/lightguardjp
> 
> Software Engineer
> Open Source Advocate
> Author of Seam Catch - Next Generation Java Exception Handling
> 
> PGP key id: 926CCFF5
> PGP key available at: keyserver.net, pgp.mit.edu
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297995

LightGuard replied:
We can't really run any arquillian tests against an OpenShift AS instance,
so it would just be there but the users wouldn't really be able to do
anything with it.

On Thu, Aug 2, 2012 at 4:51 PM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> > -            </dependencies>
> > -        </profile>
> >   +
> > -        <profile>
> > -            <!-- An optional Arquillian testing profile that executes
> >   tests in a remote JBoss AS instance -->
> > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > -            <id>arq-jbossas-remote</id>
> > -            <dependencies>
> > -                <dependency>
> > -                    <groupId>org.jboss.as</groupId>
> >   +
> >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > -                    <scope>test</scope>
> > -                </dependency>
> > -            </dependencies>
> > -        </profile>
> > -        <profile>
> 
> Well, how would this qs run on openshift?
> 
> On 2 Aug 2012, at 23:18, Jason Porter wrote:
> 
> > > -            </dependencies>
> > > -        </profile>
> > >   +
> > > -        <profile>
> > > -            <!-- An optional Arquillian testing profile that executes
> > >   tests in a remote JBoss AS instance -->
> > > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > > -            <id>arq-jbossas-remote</id>
> > > -            <dependencies>
> > > -                <dependency>
> > > -                    <groupId>org.jboss.as</groupId>
> > >   +
> > >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > > -                    <scope>test</scope>
> > > -                </dependency>
> > > -            </dependencies>
> > > -        </profile>
> > > -        <profile>
> > 
> > Ah, the openshift one. I initially didn't see the point, but we are
> > trying
> > to push openshift, aren't we?
> > 
> > On Thu, Aug 2, 2012 at 4:15 PM, Pete Muir <
> > reply@reply.github.com
> > 
> > > wrote:
> > > 
> > > > -            </dependencies>
> > > > -        </profile>
> > > >   +
> > > > -        <profile>
> > > > -            <!-- An optional Arquillian testing profile that executes
> > > >   tests in a remote JBoss AS instance -->
> > > > -            <!-- Run with: mvn clean test -Parq-jbossas-remote -->
> > > > -            <id>arq-jbossas-remote</id>
> > > > -            <dependencies>
> > > > -                <dependency>
> > > > -                    <groupId>org.jboss.as</groupId>
> > > >   +
> > > >   <artifactId>jboss-as-arquillian-container-remote</artifactId>
> > > > -                    <scope>test</scope>
> > > > -                </dependency>
> > > > -            </dependencies>
> > > > -        </profile>
> > > > -        <profile>
> > > 
> > > I meant the openshift profile :-)
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > 
> > > https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297983
> > 
> > ## 
> > 
> > Jason Porter
> > http://lightguard-jp.blogspot.com
> > http://twitter.com/lightguardjp
> > 
> > Software Engineer
> > Open Source Advocate
> > Author of Seam Catch - Next Generation Java Exception Handling
> > 
> > PGP key id: 926CCFF5
> > PGP key available at: keyserver.net, pgp.mit.edu
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297995
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1298163

## 

Jason Porter
http://lightguard-jp.blogspot.com
http://twitter.com/lightguardjp

Software Engineer
Open Source Advocate
Author of Seam Catch - Next Generation Java Exception Handling

PGP key id: 926CCFF5
PGP key available at: keyserver.net, pgp.mit.edu

pmuir replied:
Yeah, so i think remove it then ;-)

pmuir commented on cdi-portable-extension/src/main/java/org/jboss/as/quickstart/cdi/extension/CreatureExtension.java:
```diff
@@ -0,0 +1,69 @@
+    private static Document document;
+    private final Logger log = Logger.getLogger(CreatureExtension.class.getName());
+
+    static {
```
Why do this as a static?

LightGuard replied:
So it's only done once. We could move it into the method, but if people
want to be able and take and use this as a base, it makes sense to only
read and create the DOM once.

On Thu, Aug 2, 2012 at 4:12 PM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> > +import javax.xml.parsers.DocumentBuilderFactory;
> > +import javax.xml.parsers.ParserConfigurationException;
> > +
> > +import org.jboss.as.quickstart.cdi.extension.model.Creature;
> > +import org.w3c.dom.Document;
> > +import org.w3c.dom.Element;
> > +import org.xml.sax.SAXException;
> > +
> > +/**
> > - \* A simple CDI Portable Extension to "inject" values from XML into the
> >   instances of a bean.
> > - */
> >   +public class CreatureExtension implements Extension {
> > -    private static Document document;
> > -    private final Logger log =
> >   Logger.getLogger(CreatureExtension.class.getName());
> >   +
> > -    static {
> 
> Why do this as a static?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297969

## 

Jason Porter
http://lightguard-jp.blogspot.com
http://twitter.com/lightguardjp

Software Engineer
Open Source Advocate
Author of Seam Catch - Next Generation Java Exception Handling

PGP key id: 926CCFF5
PGP key available at: keyserver.net, pgp.mit.edu

pmuir replied:
well the extension is app scoped anyway.

LightGuard replied:
Oh, correct. Forgot about that. We could move that into a constructor then.

On Thu, Aug 2, 2012 at 4:15 PM, Pete Muir <
reply@reply.github.com

> wrote:
> 
> > +import javax.xml.parsers.DocumentBuilderFactory;
> > +import javax.xml.parsers.ParserConfigurationException;
> > +
> > +import org.jboss.as.quickstart.cdi.extension.model.Creature;
> > +import org.w3c.dom.Document;
> > +import org.w3c.dom.Element;
> > +import org.xml.sax.SAXException;
> > +
> > +/**
> > - \* A simple CDI Portable Extension to "inject" values from XML into the
> >   instances of a bean.
> > - */
> >   +public class CreatureExtension implements Extension {
> > -    private static Document document;
> > -    private final Logger log =
> >   Logger.getLogger(CreatureExtension.class.getName());
> >   +
> > -    static {
> 
> well the extension is app scoped anyway.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jboss-jdf/jboss-as-quickstart/pull/324/files#r1297985

## 

Jason Porter
http://lightguard-jp.blogspot.com
http://twitter.com/lightguardjp

Software Engineer
Open Source Advocate
Author of Seam Catch - Next Generation Java Exception Handling

PGP key id: 926CCFF5
PGP key available at: keyserver.net, pgp.mit.edu

pmuir commented:
Sande, I would love for you to look at this one, as I know this stuff too well!

LightGuard committed:
Removing openshift from pom, moving static block into a ctor.

sgilda commented:
I finally got a chance to look at this quickstart and it builds and runs fine! I do have a couple of comments.

After you submitted this one, James created some BOMs that allow us to define properties and remove hard-coded versions from the dependencies. I merged and closed pull 325 (https://github.com/jboss-jdf/jboss-as-quickstart/pull/325) yesterday. Would you be able to update the pom.xml to use those properties? If you're too busy, I can modify it later.

I'm a newbie and was not familiar with CDI portable extensions, so these comments may be totally off-base. Would it make sense to update the README with a brief (1 or 2 sentences) overview of CDI portable extensions and possible use cases? Also, would it be useful to give a high level description of this quickstart along with an overview of the key classes and methods in the code?  I can see where CreatureExtension implements CDI extension, parses the XML, processes the injections for the CreatureTypes, and creates the XmlBackedWrappedInjectionTarget for them. But what is purpose of the zero-length javax.enterprise.inject.spi.Extension file in the META-INF/services directory? Or would most developers know all of this? If most developers already know this, feel free to ignore my comments. :-)

LightGuard commented:
Yeah I can add that stuff (I think I am actually using the same name properties, but will need to double check). Also I may have missed something in the extension file in the META-INF/services. I'll go back and double check all this stuff and update the pull request today.

LightGuard committed:
Addressing comments made by Sande

Added a small section about PEs and what this one address in the README.
Using the new properties in the pom
Adding the class in the Extension services file (oops)

LightGuard commented:
Updated the pull request.

sgilda commented:
Nice summary! Thanks for making the fixes.
Pete, do you want to do a final review?

pmuir commented:
Merged
